File: api/planner.js
import { normalizeTitle } from './utils.js';

export function getOfficialSkillBadges(syllabus) {
  const levels = ['beginner', 'intermediate', 'advanced'];
  return levels.flatMap((levelKey) => syllabus.skill_badges?.[levelKey] || []);
}

export function skillBadgeNamesMatch(officialBadge, completedBadgeName) {
  const aliases = Array.isArray(officialBadge.aliases) ? officialBadge.aliases : [];
  const candidateNames = [officialBadge.name, ...aliases].map((name) => normalizeTitle(name));
  return candidateNames.includes(normalizeTitle(completedBadgeName));
}

export function buildArcadeGameTargets(syllabus, matchedArcadeGames) {
  const matchedIds = new Set(matchedArcadeGames.map((game) => game.id));
  return syllabus.arcade_games.map((game) => ({
    id: game.id,
    name: game.name,
    code: game.code,
    url: game.url || null,
    release_month: game.release_month || '2026-07',
    completed: matchedIds.has(game.id)
  }));
}

export function buildSkillBadgeTargets(syllabus, completedCount, completedBadges = []) {
  let remainingCompleted = completedCount;
  const completedNames = new Set(completedBadges.map((badge) => normalizeTitle(badge.name || '')).filter(Boolean));
  const shouldMatchByName = completedNames.size > 0;

  return getOfficialSkillBadges(syllabus).map((badge) => {
    const completed = shouldMatchByName
      ? completedBadges.some((completedBadge) => skillBadgeNamesMatch(badge, completedBadge.name || ''))
      : remainingCompleted > 0;
    if (!shouldMatchByName && completed) {
      remainingCompleted -= 1;
    }

    return {
      id: badge.id,
      name: badge.name,
      level: badge.level,
      url: badge.url || buildSkillBadgeCatalogUrl(badge.name),
      completed
    };
  });
}

export function buildSkillBadgeCatalogUrl(badgeName) {
  return `https://www.skills.google/catalog?skill-badge%5B%5D=skill-badge&keywords=${encodeURIComponent(badgeName)}`;
}

File: api/planner.py
from urllib.parse import quote

from .utils import normalize_title


def get_official_skill_badges(syllabus: dict) -> list[dict]:
    skill_badges = syllabus.get("skill_badges") or {}
    badges: list[dict] = []
    for level_key in ("beginner", "intermediate", "advanced"):
        badges.extend(skill_badges.get(level_key) or [])
    return badges


def skill_badge_names_match(official_badge: dict, completed_badge_name: str) -> bool:
    aliases = official_badge.get("aliases") if isinstance(official_badge.get("aliases"), list) else []
    candidate_names = [official_badge.get("name", ""), *aliases]
    return normalize_title(completed_badge_name) in {normalize_title(name) for name in candidate_names}


def build_arcade_game_targets(syllabus: dict, matched_arcade_games: list[dict]) -> list[dict]:
    matched_ids = {game.get("id") for game in matched_arcade_games}
    return [
        {
            "id": game.get("id"),
            "name": game.get("name"),
            "code": game.get("code"),
            "url": game.get("url") or None,
            "release_month": game.get("release_month") or "2026-07",
            "aliases": game.get("aliases") or [],
            "completed": game.get("id") in matched_ids,
        }
        for game in syllabus.get("arcade_games", [])
    ]


def build_skill_badge_targets(
    syllabus: dict,
    completed_count: int,
    completed_badges: list[dict] | None = None,
) -> list[dict]:
    completed_badges = completed_badges or []
    remaining_completed = completed_count
    completed_names = {normalize_title(badge.get("name", "")) for badge in completed_badges if badge.get("name")}
    should_match_by_name = len(completed_names) > 0
    targets: list[dict] = []

    for badge in get_official_skill_badges(syllabus):
        if should_match_by_name:
            completed = any(skill_badge_names_match(badge, completed_badge.get("name", "")) for completed_badge in completed_badges)
        else:
            completed = remaining_completed > 0
            if completed:
                remaining_completed -= 1

        targets.append(
            {
                "id": badge.get("id"),
                "name": badge.get("name"),
                "level": badge.get("level"),
                "url": badge.get("url") or build_skill_badge_catalog_url(badge.get("name", "")),
                "completed": completed,
            }
        )

    return targets


def build_skill_badge_catalog_url(badge_name: str) -> str:
    return f"https://www.skills.google/catalog?skill-badge%5B%5D=skill-badge&keywords={quote(badge_name)}"
